services/datamanager/datasync/sync.go

```diff
@@ -102,9 +108,14 @@ func (s *syncer) SetArbitraryFileTags(tags []string) {
 }
 
 func (s *syncer) SyncFile(path string) {
+	// Block if there the maximum number of goroutines has been hit.
+	s.syncRoutineTracker <- struct{}{}
+
 	s.backgroundWorkers.Add(1)
 	goutils.PanicCapturingGo(func() {
 		defer s.backgroundWorkers.Done()
+		// At the end, decrement the number of sync routines.
+		defer func() { <-s.syncRoutineTracker }()
```

Comment:
Getting

```
2024-01-09T11:45:42.907-0500	WARN	robot_server	impl/local_robot.go:794	resource rdk:service:data_manager/dm timed out after 1m0s during reconfigure. Update VIAM_RESOURCE_CONFIGURATION_TIMEOUT env variable to override
2024-01-09T11:45:48.600-0500	ERROR	robot_server	impl/resource_manager.go:628	error building resource	{"resource":"rdk:service:data_manager/dm","model":"rdk:builtin:builtin","error":"context deadline exceeded"}
```

for the several minutes while syncing the 300K files. The sync frequency is every 6 seconds, so it makes sense it's blocked. After syncing is complete, this goes away. I think this would be expected; curious on your thoughts for handling better.

Reply:
I'm curious about this, do you have time to explain it? It sounds like the act of syncing files blocks reconfiguration. But it's not clear to me where that blocking is happening/the purpose it would serve.

Reply:
Ah, that story makes sense now! I suspect that lock is for [reading the members](https://github.com/viamrobotics/rdk/blob/ab54ad57a235dd8e4085adfc319776e0670905ff/services/datamanager/builtin/builtin.go#L564-L569) of the `svc` object? Not that you asked, but I'd definitely try to release the lock prior to calling  `svc.sync()`. Which might look a little clumsy. Curious to see what you come up with.